bash_completion

```diff
@@ -2192,54 +2193,85 @@ else
     complete -F _cd -o nospace cd pushd
 fi
 
-# A _command_offset wrapper function for use when the offset is unknown.
-# Only intended to be used as a completion function directly associated
-# with a command, not to be invoked from within other completion functions.
-#
-_command()
-{
-    local offset i
-
-    # find actual offset, as position of the first non-option
-    offset=1
-    for ((i = 1; i <= COMP_CWORD; i++)); do
-        if [[ ${COMP_WORDS[i]} != -* ]]; then
-            offset=$i
-            break
-        fi
-    done
-    _command_offset $offset
+# Initialize regular expressions used by `_comp_command_offset__reduce_cur`.
+_comp_command_offset__initialize_regex()
+{
+    # skip if it is already initialized.
+    [[ -v _comp_command_offset__mut_initialized && $COMP_WORDBREAKS == "$_comp_command_offset__mut_initialized" ]] && return
+    _comp_command_offset__mut_initialized=$COMP_WORDBREAKS
+
+    _comp_command_offset__mut_regex_closed_prefix='^([^\"'\'']|\\.|"([^\"]|\\.)*"|'\''[^'\'']*'\'')*'
```

Comment:
> I'm not sure how long I'd have to eyeball regexps here to actually understand them and their usage fully :)

I think I can parametrize these sub-regexps by introducing the variables like `regex_single_quote_literal`.

> But one thing I don't understand yet, is what makes a literal `.` special so that it is present in here and `_comp_command_offset__mut_regex_break` below? And why it is only there with double quotes, not single ones?

`'\\.'` here becomes `\\.` in the regular expression and is treated as the pattern that matches with « one literal backslash plus following any one character ». The difference between the string literals of single quotes and double quotes is that the backslash can escape a character only in the latter string literal.

Reply:
I have refactored the regular expression generation. Actually, a large part of the non-trivial is making the regular expression `[$COMP_WORDBREAKS]` in a safe way such that `^`, `]`, `[`, and `-` would not be treated as special characters in the bracket expression `[...]`. I have isolated that part in [a new function](https://github.com/scop/bash-completion/pull/791/files#diff-a4757074ff650000804fd3eaabe9b0a9e02e33040ca5b8afd4c0275fc5f3e136R2220-R2239).

I have also "named" sub-regexps by putting them in variables and referencing them as `$name` in the main regexps. I have also added descriptions and avoided another regex hack.